src/core/server/saved_objects/migrations/integration_tests/7_13_0_unknown_types.test.ts

```diff
@@ -54,62 +57,77 @@ describe('migration v2', () => {
-      });
-
-      const client = esServer.es.getClient();
-      const { body: response } = await client.indices.getSettings(
-        { index: '.kibana_7.13.0_001' },
-        { meta: true }
-      );
-      const settings = response['.kibana_7.13.0_001'].settings as estypes.IndicesIndexSettings;
-      expect(settings.index).not.toBeUndefined();
-      expect(settings.index!.blocks?.write).not.toEqual('true');
-    }
+  describe('when `discardUnknownObjects` does not match current kibana version', () => {
+    it('fails the migration if unknown types are found in the source index', async () => {
+      // Start kibana with foo and space types disabled
+      root = createRoot('7.13.0');
+      esServer = await startES();
+      await root.preboot();
+      await root.setup();
+
+      try {
+        await root.start();
+        expect('should have thrown').toEqual('but it did not');
+      } catch (err) {
+        const errorMessage = err.message;
+
+        expect(
+          errorMessage.startsWith(
+            'Unable to complete saved object migrations for the [.kibana] index: Migration failed because some documents ' +
+              'were found which use unknown saved object types:'
+          )
+        ).toBeTruthy();
+
+        const unknownDocs = [
+          { type: 'space', id: 'space:default' },
+          { type: 'space', id: 'space:first' },
+          { type: 'space', id: 'space:second' },
+          { type: 'space', id: 'space:third' },
+          { type: 'space', id: 'space:forth' },
+          { type: 'space', id: 'space:fifth' },
+          { type: 'space', id: 'space:sixth' },
+          { type: 'foo', id: 'P2SQfHkBs3dBRGh--No5' },
+          { type: 'foo', id: 'QGSZfHkBs3dBRGh-ANoD' },
+          { type: 'foo', id: 'QWSZfHkBs3dBRGh-hNob' },
+        ];
+
+        unknownDocs.forEach(({ id, type }) => {
+          expect(errorMessage).toEqual(expect.stringContaining(`- "${id}" (type: "${type}")`));
+        });
+
+        const client = esServer.es.getClient();
+        const { body: response } = await client.indices.getSettings(
+          { index: '.kibana_7.13.0_001' },
+          { meta: true }
+        );
+        const settings = response['.kibana_7.13.0_001'].settings as IndicesIndexSettings;
+        expect(settings.index).not.toBeUndefined();
+        expect(settings.index!.blocks?.write).not.toEqual('true');
+      }
+    });
+  });
+
+  describe('when `discardUnknownObjects` matches current kibana version', () => {
+    const currentVersion = Env.createDefault(REPO_ROOT, getEnvOptions()).packageInfo.version;
+
+    it('continues with the migration even when unknown types are found in the source index', async () => {
+      // Start kibana with foo and space types disabled
+      root = createRoot(currentVersion);
+      esServer = await startES();
+      await root.preboot();
+      await root.setup();
+      await expect(root.start()).resolves.not.toThrowError();
```

Comment:
Optional: if we wanted to do e2e testing, we would query the index to make sure the objects were not indexed in the new version's index. But we have a lot of unit / integration testing covering the feature, so it's fine not to do it.

Reply:
I think it would be quite simple to assert that `.kibana` has no documents with type space/foo, so I think it's worth having this added to the test.